src/sagemaker/image.py

```diff
@@ -0,0 +1,509 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import errno
+import json
+import logging
+import os
+import platform
+import shlex
+import shutil
+import subprocess
+import sys
+import tempfile
+from subprocess import Popen, CalledProcessError
+from six.moves.urllib.parse import urlparse
+from time import sleep
+
+import yaml
+
+CONTAINER_PREFIX = "algo"
+DOCKER_COMPOSE_FILENAME = 'docker-compose.yaml'
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARNING)
+
+
+class SageMakerContainer(object):
+    """Handle the lifecycle and configuration of a local docker container execution.
+
+    This class is responsible for creating the directories and configuration files that
+    the docker containers will use for either training or serving.
+    """
+
+    def __init__(self, instance_type, instance_count, image, sagemaker_session=None):
+        """Initialize a SageMakerContainer instance
+
+        It uses a :class:`sagemaker.session.Session` for general interaction with user configuration
+        such as getting the default sagemaker S3 bucket. However this class does not call any of the
+        SageMaker APIs.
+
+        Args:
+            instance_type (str): The instance type to use. Either 'local' or 'local_gpu'
+            instance_count (int): The number of instances to create.
+            image (str): docker image to use.
+            sagemaker_session (sagemaker.session.Session): a sagemaker session to use when interacting
+                with SageMaker.
+        """
+        from sagemaker.local_session import LocalSession
+        self.sagemaker_session = sagemaker_session or LocalSession()
+        self.instance_type = instance_type
+        self.instance_count = instance_count
+        self.image = image
+        self.hosts = ['{}-{}'.format(CONTAINER_PREFIX, i) for i in range(1, self.instance_count + 1)]
+        self.container_root = None
```

Comment:
container_root and container could do with some developer documentation explaining what they are